src/ring_theory/power_basis.lean

```diff
@@ -239,98 +242,41 @@ noncomputable def adjoin.power_basis [algebra K S] {x : S} (hx : _root_.is_integ
-    rw inv_eq_of_root_of_coeff_zero_ne_zero
-      (minimal_polynomial.aeval hx) (minimal_polynomial.coeff_zero_ne_zero hx hx0),
-    use - (C ((minimal_polynomial hx).coeff 0)⁻¹) * (minimal_polynomial hx).div_X.comp fx,
-    rw aeval_def at x_eq,
-    rw [div_eq_inv_mul, alg_hom.map_mul, alg_hom.map_neg, aeval_C, neg_mul_eq_neg_mul,
-        ring_hom.map_inv, aeval_def, aeval_def, eval₂_comp, ← x_eq] },
-  { rintros x y ⟨fx, rfl⟩ ⟨fy, rfl⟩,
-    exact ⟨fx * fy, (ring_hom.map_mul _ _ _).symm⟩ },
-end
-
-lemma power_basis_is_basis (alg : algebra.is_algebraic K L) {x : L} (hx : is_integral K x) :
-  is_basis K (λ (i : fin (minimal_polynomial hx).nat_degree), (adjoin_simple.gen K x ^ (i : ℕ))) :=
-begin
-  have hx' : is_integral K (adjoin_simple.gen K x),
-  { apply (is_integral_algebra_map_iff (algebra_map K⟮x⟯ L).injective).mp,
-    convert hx,
-    apply_instance },
-  have minpoly_eq :=
-    minimal_polynomial.eq_of_algebra_map_eq ((algebra_map K⟮x⟯ L).injective) hx' hx,
-  refine ⟨_, eq_top_iff.mpr _⟩,
-  { have := algebra.linear_independent_power_basis hx',
-    rwa minpoly_eq at this,
-    refl },
-  { rintros ⟨y, hy⟩ _,
-    have := algebra.mem_span_power_basis hx',
-    rw minpoly_eq at this,
-    apply this,
-    { obtain ⟨f, rfl⟩ := adjoin_simple.exists_eq_aeval_gen alg hy,
-      use f,
-      ext,
-      exact (is_scalar_tower.algebra_map_aeval K K⟮x⟯ L (adjoin_simple.gen K x) _).symm },
-    { refl } }
-end
-
-/-- The power basis `1, x, ..., x ^ (d - 1)` for `K⟮x⟯`,
-where `d` is the degree of the minimal polynomial of `x`. -/
-noncomputable def adjoin.power_basis (alg : algebra.is_algebraic K L) {x : L} :
-  power_basis K K⟮x⟯ :=
-{ gen := adjoin_simple.gen K x,
-  dim := (minimal_polynomial ((is_algebraic_iff_is_integral K).mp (alg x))).nat_degree,
-  is_basis := power_basis_is_basis alg _ }
-
-end intermediate_field
-
 namespace adjoin_root
 
-lemma power_basis_is_basis {f : polynomial K} (f_irr : irreducible f) :
-  is_basis K (λ (i : fin f.nat_degree), (root f ^ (i : ℕ))) :=
+variables {f : polynomial K}
+
+lemma power_basis_is_basis (hf : f ≠ 0) : is_basis K (λ (i : fin f.nat_degree), (root f ^ i.val)) :=
 begin
   set f' := f * C (f.leading_coeff⁻¹) with f'_def,
-  have hC : f.leading_coeff ≠ 0,
-  { simpa only [ne.def, C_eq_zero, inv_eq_zero, leading_coeff_eq_zero] using f_irr.ne_zero},
   have deg_f' : f'.nat_degree = f.nat_degree,
-  { rw [nat_degree_mul f_irr.ne_zero (mt _ hC), nat_degree_C, add_zero],
-    simp },
-  have f'_monic : monic f' := monic_mul_leading_coeff_inv f_irr.ne_zero,
-  have f'_irr : irreducible f',
-  { apply irreducible_of_associated _ f_irr,
-    refine ⟨⟨C f.leading_coeff⁻¹, C f.leading_coeff, _, _⟩, _⟩;
-      simp only [← C.map_mul, mul_inv_cancel hC, inv_mul_cancel hC, C.map_one, units.coe_mk] },
+  { rw [nat_degree_mul hf, nat_degree_C, add_zero],
+    rw [ne.def, C_eq_zero, inv_eq_zero, leading_coeff_eq_zero],
+    exact hf },
```

Comment:
Let's make clear(er) that the second `rw` solves a goal introduced by the first `rw`.
```suggestion
  { rw [nat_degree_mul hf, nat_degree_C, add_zero],
    { rwa [ne.def, C_eq_zero, inv_eq_zero, leading_coeff_eq_zero] } },
```